core/embed/rust/src/protobuf/decode.rs

```diff
@@ -0,0 +1,309 @@
+        };
+
+        let obj = decoder.message_from_stream(stream, def.msg())?;
+        Ok(obj)
+    })
+}
+
+pub struct Decoder {
+    pub enable_experimental: bool,
+}
+
+impl Decoder {
+    /// Create a new message instance and decode `stream` into it, handling the
+    /// default and required fields correctly.
+    pub fn message_from_stream(
+        &self,
+        stream: &mut InputStream,
+        msg: &MsgDef,
+    ) -> Result<Obj, Error> {
+        let mut dict = self.empty_message(msg);
+        // SAFETY: We assume that `obj` is not alised here.
+        let map = unsafe { dict.as_mut() }.map_mut();
+        self.decode_fields_into(stream, msg, map)?;
+        self.decode_defaults_into(msg, map)?;
+        self.assign_required_into(msg, map)?;
+        Ok(dict.into())
+    }
+
+    /// Create a new message instance and fill it from `values`, handling the
+    /// default and required fields correctly.
+    pub fn message_from_values(&self, values: &Map, msg: &MsgDef) -> Result<Obj, Error> {
+        let mut obj = self.empty_message(msg);
+        // SAFETY: We assume that `obj` is not alised here.
+        let map = unsafe { obj.as_mut() }.map_mut();
+        for elem in values.elems() {
+            map.set(elem.key, elem.value);
+        }
+        self.decode_defaults_into(msg, map)?;
+        self.assign_required_into(msg, map)?;
+        Ok(obj.into())
+    }
+
+    /// Allocate the backing message object with enough pre-allocated space for
+    /// all fields.
+    pub fn empty_message(&self, msg: &MsgDef) -> Gc<MsgObj> {
+        MsgObj::alloc_with_capacity(msg.fields.len(), msg)
+    }
+
+    /// Decode message fields one-by-one from the input stream, assigning them
+    /// into `map`.
+    pub fn decode_fields_into(
+        &self,
+        stream: &mut InputStream,
+        msg: &MsgDef,
+        map: &mut Map,
+    ) -> Result<(), Error> {
+        // Loop, trying to read the field key that contains the tag and primitive value
+        // type. If we fail to read the key, we are at the end of the stream.
+        while let Ok(field_key) = stream.read_uvarint() {
+            let field_tag = u8::try_from(field_key >> 3)?;
+            let prim_type = u8::try_from(field_key & 7)?;
+
+            match msg.field(field_tag) {
+                Some(field) => {
+                    let field_value = self.decode_field(stream, field)?;
+                    let field_name = Qstr::from(field.name);
+                    if field.is_repeated() {
+                        // Repeated field, values are stored in a list. First, look up the list
+                        // object. If it exists, append to it. If it doesn't, create a new list with
+                        // this field's value and assign it.
+                        if let Ok(obj) = map.get(field_name) {
+                            let mut list = Gc::<List>::try_from(obj)?;
+                            // SAFETY: We assume that `list` is not aliased here.
```

Comment:
I'll just change the comment :)

> my natural way of thinking about it would make this function return the map, or perhaps the dict, but something tells me that could get awkward in Rust?

I see absolutely no reason to return the map.

Reply:
Oh you mean create it inside the fn and return it? We could do that I guess, but here it has no practical benefit I think.

Reply:
> seems this function should not be `pub` in any case

also this? or is there a reason to have it public?